service/history/historyEngine.go

```diff
@@ -183,7 +183,7 @@ var (
 	// ErrQueryEnteredInvalidState is error indicating query entered invalid state
 	ErrQueryEnteredInvalidState = &workflow.BadRequestError{Message: "query entered invalid state, this should be impossible"}
 	// ErrQueryWorkflowBeforeFirstDecision is error indicating that query was attempted before first decision task completed
-	ErrQueryWorkflowBeforeFirstDecision = &workflow.BadRequestError{Message: "workflow must handle at least one decision task before it can be queried"}
+	ErrQueryWorkflowBeforeFirstDecision = errors.New("workflow must handle at least one decision task before it can be queried")
```

Comment:
How about QueryFailedError?

Reply:
We can make it queryFailedError but there is a work item here in that we over use this error type. We need an error type that indicates the users code is the reason the query failed. I was planning to use queryFailedError to indicate this. But if not we should consider introducing a new error type like "QueryWorkflowFailedError" (the naming here is hard).

Reply:
Right now just by looking at the error type its hard to determine if the queryFailure was because of users code or not.

Reply:
I'm fine with creating a new error type. The concern here is that user is not able to assert on this general error and do backoff/retry.

Reply:
Sounds good I will create a new error type.

Reply:
Maybe this should not belong here. But we should create two generic types of errors. Retryable and NonRetryable.